bottomless/src/replicator.rs

```diff
@@ -607,59 +633,83 @@ impl Replicator {
+        let mut db_file = File::open(&self.db_path).await?;
+        let change_counter = Self::read_change_counter(&mut db_file).await?;
+        let gen = self.generation.load_full();
+        let snapshot_req = client.put_object().bucket(self.bucket.clone()).key(format!(
+            "{}-{}/db.{}",
+            self.db_name, gen, self.use_compression
+        ));
 
         /* FIXME: we can't rely on the change counter in WAL mode:
          ** "In WAL mode, changes to the database are detected using the wal-index and
          ** so the change counter is not needed. Hence, the change counter might not be
          ** incremented on each transaction in WAL mode."
          ** Instead, we need to consult WAL checksums.
          */
-        let change_counter_key = format!("{}-{}/.changecounter", self.db_name, self.generation);
-        self.client
+        let change_counter_key = format!("{}-{}/.changecounter", self.db_name, gen);
+        let change_counter_req = self
+            .client
             .put_object()
             .bucket(&self.bucket)
             .key(change_counter_key)
-            .body(ByteStream::from(Bytes::copy_from_slice(&change_counter)))
-            .send()
-            .await?;
-        tracing::debug!("Main db snapshot complete in {:?}", Instant::now() - start);
-        Ok(true)
+            .body(ByteStream::from(Bytes::copy_from_slice(
+                change_counter.as_ref(),
+            )));
+        let snapshot_notifier = self.snapshot_notifier.clone();
+        let compression = self.use_compression;
+        let handle = tokio::spawn(async move {
+            let start = Instant::now();
+            let body = match Self::maybe_compress_main_db_file(db_file, compression).await {
+                Ok(file) => file,
+                Err(e) => {
+                    tracing::error!("Failed to compress db file (generation {}): {}", gen, e);
+                    let _ = snapshot_notifier.send(Err(e));
+                    return;
+                }
+            };
+            let mut result = snapshot_req.body(body).send().await;
+            if let Err(e) = result {
+                tracing::error!("Failed to upload snapshot for generation {}: {:?}", gen, e);
+                let _ = snapshot_notifier.send(Err(e.into()));
+                return;
+            }
+            result = change_counter_req.send().await;
+            if let Err(e) = result {
+                tracing::error!(
+                    "Failed to upload change counter for generation {}: {:?}",
+                    gen,
+                    e
+                );
+                let _ = snapshot_notifier.send(Err(e.into()));
+                return;
+            }
+            let _ = snapshot_notifier.send(Ok(Some(gen)));
+            let elapsed = Instant::now() - start;
+            tracing::debug!("Snapshot upload finished (took {:?})", elapsed);
+            let _ = tokio::fs::remove_file(format!("db.{}", compression)).await;
```

Comment:
This file should probably be removed before every return below line 687.

Reply:
It may be potentially useful to keep this file around. Whenever we try to restore DB, one of the first steps is to check if there are missing files, which have been not uploaded to S3 in the previous process session, then try to re upload them again. This way we may potentially overcome potential S3 issues.